src/Pos/Client/Txp/History.hs

```diff
@@ -78,84 +80,78 @@ data TxHistoryAnswer = TxHistoryAnswer
 ----------------------------------------------------------------------
 
 -- | Check if given 'Address' is one of the receivers of 'Tx'
-hasReceiver :: Tx -> Address -> Bool
-hasReceiver UnsafeTx {..} addr = any ((== addr) . txOutAddress) _txOutputs
+hasReceiver :: Tx -> Set Address -> Set Address
+hasReceiver UnsafeTx {..} addrs =
+    addrs `S.intersection` (S.fromList $ toList (txOutAddress <$> _txOutputs))
 
 -- | Given some 'Utxo', check if given 'Address' is one of the senders of 'Tx'
-hasSender :: MonadUtxoRead m => Tx -> Address -> m Bool
-hasSender UnsafeTx {..} addr = anyM hasCorrespondingOutput $ toList _txInputs
-  where hasCorrespondingOutput txIn =
-            fmap toBool $ ((== addr) . txOutAddress . toaOut) <<$>> utxoGet txIn
-        toBool Nothing  = False
-        toBool (Just b) = b
+hasSender :: MonadUtxoRead m => Tx -> Set Address -> m [Address]
+hasSender UnsafeTx {..} addrs =
+  catMaybes <$> mapM fetchWithCorrespondingOutput (toList _txInputs)
+  where fetchWithCorrespondingOutput txIn = runMaybeT $ do
+          inAddr <- MaybeT $ txOutAddress . toaOut <<$>> utxoGet txIn
+          guard (inAddr `S.member` addrs)
+          return inAddr
 
 -- | Datatype for returning info about tx history
 data TxHistoryEntry = THEntry
     { _thTxId       :: !TxId
     , _thTx         :: !Tx
     , _thIsOutput   :: !Bool
     , _thDifficulty :: !(Maybe ChainDifficulty)
+    , _thInputAddr  :: !Address
     } deriving (Show, Eq, Generic)
 
 makeLenses ''TxHistoryEntry
 
 -- | Type of monad used to deduce history
 type TxSelectorT m = UtxoStateT (MaybeT m)
 
--- | Select transactions related to given address. `Bool` indicates
+-- | Select transactions related to given addresses. `Bool` indicates
 -- whether the transaction is outgoing (i. e. is sent from given address)
```

Comment:
I guess. By the way
> because it's signed with one key
As I understand, it will be signed with one key even if its many-to-many transaction: with your primary key. Am I right @flyingleafe  ?